lib/Ansi/Ansi.mli

```diff
@@ -0,0 +1,182 @@
+(** This module defines the escape sequence data type.
+
+    Every value of this type is one styling ([`Bold],
+    [`Reverse], [`Foreground] {!Color.red}) or the composition
+    of two of them ([`Bold]{{!(&)}[&]}[`Reverse]), which can be chained
+    ([`Dim] {{!(&)}[&]} [`Italic] {{!(&)}[&]} [`Blink] {{!(&)}[&]} [`Background] {!Color.magenta}).
+
+    {[
+    open Ansifmt
+
+    let chesnay = `Italic
+    let rocquencout = Ansi.(`Bold & `Reverse & `Foreground Color.red)
+
+    (* compose to infinity! *)
+    let inria = Ansi.(chesnay & rocquencourt)
+    ]}
+
+    {i [chesnay], [rocquencourt] and [inria] all have the same
+    type {!t}.}
+
+    In a composition chain, if two stylings are overlapping
+    (for example, [`Foreground Color.blue & `Foreground Color.red]),
+    the rightmost takes precedence.
+    *)
+
+(** Represents an ANSI SGR escape sequence. *)
+type t =
+  [ `Bold
+  | `Dim
+  | `Italic
+  | `Underline
+  | `Blink
+  | `Reverse
+  | `Foreground of Color.t
+  | `Background of Color.t
+  | `Composed of t * t
+  ]
+
+(** [compose left right] combines the [left] and right
+    sequences into one.
+
+    See also: {{!(&)}[( & )]} *)
+val compose : t -> t -> t
+
+(** [to_attributes ansi] produces an attribute record of the
+    [ansi] escape sequence. *)
+val to_attributes : t -> Attributes.t
+
+(** [to_cancelling_attributes ansi] produces an attribute
+    record of the escape sequence that would cancel the effects
+    of [ansi]. *)
+val to_cancelling_attributes : t -> Attributes.t
+
+(** [serialize ansi] produces a serialized representation of
+    the [ansi] escape sequence.
+
+    {b Example}
+
+    {[
+    open Ansifmt
+
+    let styling = Ansi.(`Dim & `Background (`Rgb (0, 0, 0)))
+    let () = Printf.printf "%s\n" (Ansi.serialize styling)
+    ]}
+
+    {i [dim & background(rgb(0, 0, 0))] will be printed.}
+
+    See also: {!deserialize}.
+    *)
+val serialize : t -> string
+
+(** [deserialize string] produces an escape sequence from a
+    serialization [string]. If it fails to parse, returns
+    [None].
+    
+    {b Example}
+
+    {[
+    open Ansifmt
+
+    let serialized = "blink & BOLD & foreground(rgb(255, 0, 0))"
+
+    let () =
+    match Ansi.deserialize serialized with
+    | None -> Printf.eprintf "oh nothing don't worry\n"
+    | Some ansi -> 
+        Printf.printf "%sHACKED!\n" (Ansi.show ansi)
+    ;;
+    ]}
+
+    {i A blinking bold [HACKED!] will be printed in red.}
+
+    See also: {!serialize}.
+    *)
+val deserialize : string -> t option
```

Comment:
Suggestion: Consider a `result` type if you can provide meaningful error messages, otherwise `Option` is a good fallback.